apisix/plugins/api-breaker.lua

```diff
@@ -0,0 +1,221 @@
+    error("failed to get ngx.shared dict when load plugin " .. plugin_name)
+end
+
+
+local schema = {
+    type = "object",
+    properties = {
+        unhealthy_response_code = {
+            type = "integer",
+            minimum = 200,
+            maximum = 599,
+        },
+        unhealthy = {
+            type = "object",
+            properties = {
+                http_statuses = {
+                    type = "array",
+                    minItems = 1,
+                    items = {
+                        type = "integer",
+                        minimum = 500,
+                        maximum = 599,
+                    },
+                    uniqueItems = true,
+                    default = {500}
+                },
+                failures = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 1,
+                }
+            }
+        },
+        healthy = {
+            type = "object",
+            properties = {
+                http_statuses = {
+                    type = "array",
+                    minItems = 1,
+                    items = {
+                        type = "integer",
+                        minimum = 200,
+                        maximum = 499,
+                    },
+                    uniqueItems = true,
+                    default = {200, 206}
+                },
+                successes = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 1,
+                }
+            }
+        }
+    },
+    required = {"unhealthy_response_code", "unhealthy", "healthy"},
+}
+
+
+local function is_unhealthy(unhealthy_status, upstream_status)
+    for _, unhealthy in ipairs(unhealthy_status) do
+        if unhealthy == upstream_status then
+            return true
+        end
+    end
+
+    return false
+end
+
+
+local function is_healthy(healthy_status, upstream_status)
+    for _, healthy in ipairs(healthy_status) do
+        if healthy == upstream_status then
+            return true
+        end
+    end
+
+    return false
+end
+
+
+local function healthy_cache_key(ctx)
+    return "healthy-" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local function unhealthy_cache_key(ctx)
+    return "unhealthy-" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local function unhealthy_lastime_cache_key(ctx)
+    return "unhealthy-lastime" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local _M = {
+    version = 0.1,
+    name = plugin_name,
+    priority = 1005,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+function _M.access(conf, ctx)
+    local unhealthy_val, err = shared_buffer:get(unhealthy_cache_key(ctx))
+    if err then
+        core.log.error("failed to get unhealthy_cache_key in ngx.shared:", err)
```

Comment:
do we need error level log?
And we need to log the value of `unhealthy_cache_key(ctx)` for debug.

Reply:
updated to warn